light/beacon/sync_committee.go

```diff
@@ -0,0 +1,547 @@
+			sc, _ := v.(syncCommittee)
+			return sc
+		}
+		if sc := s.GetSerializedSyncCommittee(period, committeeRoot); sc != nil {
+			c := s.sigVerifier.deserializeSyncCommittee(sc)
+			s.syncCommitteeCache.Add(key, c)
+			return c
+		} else {
+			log.Error("Missing serialized sync committee", "period", period, "committeeRoot", committeeRoot)
+		}
+	}
+	return nil
+}
+
+// EnforceForksAndConstraints rolls back committee updates that do not match the tracker's forks
+// and constraints and also starts new requests if possible (tracker mutex locked)
+func (s *SyncCommitteeTracker) EnforceForksAndConstraints() {
+	s.lock.Lock()
+	s.enforceForksAndConstraints()
+	s.lock.Unlock()
+}
+
+// enforceForksAndConstraints rolls back committee updates that do not match the tracker's forks
+// and constraints and also starts new requests if possible (tracker mutex expected)
+func (s *SyncCommitteeTracker) enforceForksAndConstraints() {
+	if !s.genesisInit || !s.chainInit {
+		return
+	}
+	s.committeeRootCache.Purge()
+	for s.nextPeriod > s.firstPeriod {
+		if update := s.GetBestUpdate(s.nextPeriod - 1); update == nil || s.verifyUpdate(update) { // check constraints and signature
+			if update == nil {
+				log.Error("Sync committee update missing", "period", s.nextPeriod-1)
+			}
+			break
+		}
+		s.nextPeriod--
+		s.deleteBestUpdate(s.nextPeriod)
+	}
+	if s.nextPeriod == s.firstPeriod {
+		if root, matchAll := s.constraints.CommitteeRoot(s.firstPeriod); matchAll || s.getSyncCommitteeRoot(s.firstPeriod) != root || s.getSyncCommittee(s.firstPeriod) == nil {
+			s.nextPeriod, s.firstPeriod, s.chainInit = 0, 0, false
+		}
+	}
+
+	s.retrySyncAllPeers()
+}
+
+// checkConstraints checks whether the signed headers of the given committee update is
+// on the right fork and the proven NextSyncCommitteeRoot matches the update chain constraints.
+func (s *SyncCommitteeTracker) checkConstraints(update *LightClientUpdate) bool {
+	if !s.genesisInit {
+		log.Error("SyncCommitteeTracker not initialized")
+		return false
+	}
+	root, matchAll := s.constraints.CommitteeRoot(uint64(update.Header.Slot)>>13 + 1)
+	return matchAll || root == update.NextSyncCommitteeRoot
+}
+
+// LightClientUpdate is a proof of the next sync committee root based on a header signed by the sync committee
+// of the given period. Optionally the update can prove quasi-finality by the signed header referring to a previous,
+// finalized header from the same period, and the finalized header referring to the next sync committee root.
+type LightClientUpdate struct {
+	Header                  Header
+	NextSyncCommitteeRoot   common.Hash
+	NextSyncCommitteeBranch MerkleValues
+	FinalizedHeader         Header
+	FinalityBranch          MerkleValues
+	SyncCommitteeBits       []byte
+	SyncCommitteeSignature  []byte
+	score                   UpdateScore
+}
+
+// Validate verifies the validity of the update
+func (update *LightClientUpdate) Validate() error {
+	if update.Header.Slot&0x1fff == 0x1fff {
```

Comment:
We can define the function of header `lastSlotOfPeriod`, it will make code more readable.

Reply:
Actually this whole solution was ugly, now it's not needed at all because I pass the assumed `signatureSlot` separately to the sig verifier so I can just assume that updates are always signed by the same committee that the header belongs to (that is a validity condition of the header, the sig verification will fail if it's not true).